### PR TITLE
Add ensure-name-substring function

### PR DESCRIPTION
**Pull request by mengqiy**
This supersedes #168

fixes: https://github.com/GoogleContainerTools/kpt/issues/1474
fixes: https://github.com/GoogleContainerTools/kpt/issues/1475

### CODE REVIEW

File: functions/go/ensure-name-segment/ensure_name_segment.go [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+
+type EnsureNameSegment struct {
+	yaml.ResourceMeta `json:",inline" yaml:",inline"`
+	// Segment is the desired name segment.
+	Segment string `json:"segment" yaml:"segment"`
+	// ActionOnNotFound controls the desired action when the desired segment is not found in the name.
+	// If not specified, prepend will be the default.
+	ActionOnNotFound ActionMode `json:"actionOnNotFound,omitempty" yaml:"actionOnNotFound,omitempty"`
+
+	FieldSpecs []types.FieldSpec `json:"fieldSpecs,omitempty" yaml:"fieldSpecs,omitempty"`
```
Comment (droot) [inline functions/go/ensure-name-segment/ensure_name_segment.go:33]: I am guessing for most common use cases name in the metadata of a KRM resource, users won't have to specify this. Right ?

File: functions/go/ensure-name-segment/ensure_name_segment.go [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+	fnConfigKind       = "EnsureNameSegment"
+)
+
+type EnsureNameSegment struct {
+	yaml.ResourceMeta `json:",inline" yaml:",inline"`
+	// Segment is the desired name segment.
+	Segment string `json:"segment" yaml:"segment"`
+	// ActionOnNotFound controls the desired action when the desired segment is not found in the name.
+	// If not specified, prepend will be the default.
+	ActionOnNotFound ActionMode `json:"actionOnNotFound,omitempty" yaml:"actionOnNotFound,omitempty"`
```
Comment (droot) [inline functions/go/ensure-name-segment/ensure_name_segment.go:31]: `EditMode` might be more intuitive.

File: functions/go/ensure-name-segment/ensure_name_segment.go [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+	fnConfigGroup      = "fn.kpt.dev"
+	fnConfigVersion    = "v1alpha1"
+	fnConfigAPIVersion = fnConfigGroup + "/" + fnConfigVersion
+	fnConfigKind       = "EnsureNameSegment"
+)
+
+type EnsureNameSegment struct {
+	yaml.ResourceMeta `json:",inline" yaml:",inline"`
+	// Segment is the desired name segment.
+	Segment string `json:"segment" yaml:"segment"`
```
Comment (droot) [inline functions/go/ensure-name-segment/ensure_name_segment.go:28]: Sorry.. some bikeshedding :) 

Does `SubString` reflect the usage better ?

and calling `EnsureNameSegment` --> `EnsureNameSubString` reflect the usage better ?

File: functions/go/ensure-name-segment/ensure_name_segment.go [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+		// only one entry: "metadata/name" with no GVK
+		for _, fs := range ens.FieldSpecs {
+			// TODO: this is redundant to filter (but needed for now)
+			if !id.IsSelected(&fs.Gvk) {
+				continue
+			}
+
+			// Idempotency check: if the segment is already part of the name, we
+			// don't need to do anything.
+			contain, err := resourceContainsSegment(r, ens.Segment, fs)
```
Comment (droot) [inline functions/go/ensure-name-segment/ensure_name_segment.go:72]: s/contain/contains or `hasSegment`

File: functions/go/ensure-name-segment/ensure_name_segment.go [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+				return err
+			}
+			if contain {
+				continue
+			}
+
+			if isNameChange(&fs) {
+				// If we are changing "metadata/name", we tracks the original
+				// name and the prefix or suffix being added.
+				r.StorePreviousId()
```
Comment (droot) [inline functions/go/ensure-name-segment/ensure_name_segment.go:83]: So its reusing kustomize's implementation. That gives us good confidence in the functionality.

File: examples/ensure-name-segment/advanced/README.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+kpt fn run advanced
+```
+
+## Expected result
+
+Check all resources have `prod-` in their names and the field `.spec.name` in
+`MyResource` also got updated.
+
+```sh
+kpt cfg cat advanced
```
Comment (droot) [inline examples/ensure-name-segment/advanced/README.md:41]: Do we have `cfg` command group ?

**Review by droot — COMMENTED** [review]
I only have some `bikeshedding` level comments :) mostly about the naming.

IMO, we should release this function with the intention to get some feedback internally for the CUJs that involves prefixing/suffixing from our team. We should have some alternative names in mind and behavior to propose in response to the feedback from the team and then go ahead for the final push.

**Comment by droot** [comment]
One comment for the docs: I would suggest indicating something like `Alpha` or `Experimental` for this function to indicate that we are seeking feedback actively about the UX for this function.

File: functions/go/ensure-name-segment/ensure_name_segment.go (hunk 1 above)
Comment (mengqiy) [reply functions/go/ensure-name-segment/ensure_name_segment.go:33]: Right, for common use cases, users don't need to specify it and the default ones should do the job for them.
I imagine users will use ConfigMap for common use cases.

File: functions/go/ensure-name-segment/ensure_name_segment.go (hunk 2 above)
Comment (mengqiy) [reply functions/go/ensure-name-segment/ensure_name_segment.go:31]: I'm a little confused here. Are you suggesting  changing `actionOnNotFound` to `EditMode` or change `ActionMode`?

File: functions/go/ensure-name-segment/ensure_name_segment.go (hunk 3 above)
Comment (mengqiy) [reply functions/go/ensure-name-segment/ensure_name_segment.go:28]: `Substring` sounds better.

File: examples/ensure-name-segment/advanced/README.md (hunk 6 above)
Comment (mengqiy) [reply examples/ensure-name-segment/advanced/README.md:41]: We are currently using released kpt in our tests.
We will do bulk update to all examples when kpt approaching v1 release.

File: functions/go/ensure-name-segment/ensure_name_segment.go (hunk 5 above)
Comment (mengqiy) [reply functions/go/ensure-name-segment/ensure_name_segment.go:83]: Yes, we try to reuse kustomize's implementation as much as possible.

File: functions/go/ensure-name-segment/README.md [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+Ensure a name segment is part of the name.
+
+<!--mdtogo-->
+
+### Synopsis
+
+<!--mdtogo:Long-->
+
+If the desired name segment is already part of the name, it takes no actions.
```
Comment (Shell32-Natsu) [inline functions/go/ensure-name-segment/README.md:15]: What will happen if user want to add suffix `-deploy` to name `resource-deployment`? I suppose the suffix will not be added?

File: functions/go/ensure-name-segment/ensure_name_segment.go (hunk 2 above)
Comment (droot) [reply functions/go/ensure-name-segment/ensure_name_segment.go:31]: oh.. change `actionOnNotFound` to `editMode`, the user-facing part.

File: functions/go/ensure-name-segment/README.md (hunk 7 above)
Comment (mengqiy) [reply functions/go/ensure-name-segment/README.md:15]: It won't be added in this case. This is going to be the limitation.

Comment (Shell32-Natsu) [reply functions/go/ensure-name-segment/README.md]: Probably we should list these limitations in the readme

**Comment by mengqiy** [comment]
@droot @Shell32-Natsu PTAL

File: functions/go/ensure-name-segment/README.md (hunk 7 above)
Comment (mengqiy) [reply functions/go/ensure-name-segment/README.md]: Yes, I added it in https://github.com/GoogleContainerTools/kpt-functions-catalog/pull/226/files#diff-3c677f0bba9846eff1255ea8e3d7773948e2d4832c4b4439e7e89745ad365c81R18

**Comment by mengqiy** [comment]
>One comment for the docs: I would suggest indicating something like Alpha or Experimental for this function to indicate that we are seeking feedback actively about the UX for this function.

I guess we should mark it as experimental in both usage doc and help text.

File: examples/ensure-name-segment/prefix/README.md [hunk 8]
```diff
@@ -0,0 +1,36 @@
+# ensure-name-substring: Prefix Example
+
```
Comment (droot) [inline examples/ensure-name-segment/prefix/README.md:2]: I think adding a note about it being experimental here for all the docs (`Help`, `Prefix`, `Suffix`, `Advanced`)

```
Note: This is an alpha function and we are actively seeking feedback on the function config syntax and behavior. If you have suggestion or feedback, please file a issue [here](https://github.com/GoogleContainerTools/kpt/issues/new/choose).
```

**Review by droot — APPROVED** [review]
minor comments, and then ship it.

File: examples/ensure-name-segment/advanced/README.md (hunk 6 above)
Comment (mengqiy) [reply examples/ensure-name-segment/advanced/README.md:45]: It is being tracked in https://github.com/GoogleContainerTools/kpt/issues/1548

**Review by droot — APPROVED** [review]

File: functions/go/ensure-name-substring/README.md [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+Note: If the original name of a resource happens to contain the desired
+substring, the desired substring will not be added again. Users need to ensure
+the name collisions don't happen.
+
+To configure it using a ConfigMap, only one key-value pair is allowed in `data`
+field. The key must be one of `prepend` and `append`, and the value is the
+desired name substring. 
+
+```yaml
+prepend|append: Desired name substring
```
Comment (Shell32-Natsu) [inline functions/go/ensure-name-substring/README.md:31]: Can we have an option to restrict the position of substring? For example I only want to ensure substring `dev-` in the beginning of the name.

File: functions/go/ensure-name-substring/ensure_name_substring.go [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+		return fmt.Errorf("only 1 entry is allowed in the ConfigMap, but got: %d", len(cm.Data))
+	}
+	for k, v := range cm.Data {
+		switch k {
+		case string(Prepend):
+			ens.EditMode = Prepend
+		case string(Append):
+			ens.EditMode = Append
+		default:
+			return fmt.Errorf("editMode")
```
Comment (Shell32-Natsu) [inline functions/go/ensure-name-substring/ensure_name_substring.go:154]: Can we have a better error message here?

File: functions/go/ensure-name-substring/ensure_name_substring.go [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+func resourceContainsSubstring(r *resource.Resource, substring string, fs types.FieldSpec) (bool, error) {
+	if isNameChange(&fs) {
+		// Idempotency check: if the substring is already part of the name, we
+		// don't need to do anything.
+		return strings.Contains(r.GetName(), substring), nil
+	}
+
+	rn, err := yaml.FromMap(r.Map())
+	if err != nil {
+		return false, err
```
Comment (Shell32-Natsu) [inline functions/go/ensure-name-substring/ensure_name_substring.go:191]: It's better to provide more error information.

File: functions/go/ensure-name-substring/README.md (hunk 9 above)
Comment (mengqiy) [reply functions/go/ensure-name-substring/README.md:31]: I guess we can do that in another function "ensure-name-prefix" which is approach 1 in go/kpt-fn-set-name-prefix-simplified

Comment (Shell32-Natsu) [reply functions/go/ensure-name-substring/README.md:31]: ack

**Review by Shell32-Natsu — APPROVED** [review]

**Pull request merged**